src/core/model_config_utils.cc

```diff
@@ -382,48 +385,233 @@ ValidateModelConfig(
+        " whose platform is not ", kEnsemblePlatform);
+  }
+  if (config.instance_group().size() != 0) {
+    return tensorflow::errors::InvalidArgument(
+        "instance group should not be specified for ensemble ", config.name());
+  }
+  if (config.has_optimization()) {
+    return tensorflow::errors::InvalidArgument(
+        "optimization should not be specified for ensemble ", config.name());
+  }
+
+  // Make sure step is not empty and all fields are set
+  if (config.ensemble_scheduling().step_size() == 0) {
+    return tensorflow::errors::InvalidArgument(
+        "must specify 'step' for ensemble ", config.name());
+  }
+
+  std::unordered_map<std::string, EnsembleTensor> tensors;
+
+  TF_RETURN_IF_ERROR(BuildEnsembleGraph(config, tensors));
+
+  // check data flow
+  std::deque<EnsembleTensor*> ready_queue;
+  for (const auto& input : config.input()) {
+    auto it = tensors.find(input.name());
+    if (it == tensors.end()) {
+      return tensorflow::errors::InvalidArgument(
+          "must map ensemble input ", input.name(), " for ensemble ",
+          config.name());
+    }
+    it->second.ready = true;
+    ready_queue.push_back(&(it->second));
+  }
+  while (!ready_queue.empty()) {
+    auto& ready_node = ready_queue.front();
+    for (auto& next_node : ready_node->next_nodes) {
+      if (next_node->ready) {
+        continue;
+      }
+      bool next_node_ready = true;
+      for (auto& prev_node : next_node->prev_nodes) {
+        if (!prev_node->ready) {
+          next_node_ready = false;
+          break;
+        }
+      }
+      next_node->ready = next_node_ready;
+      if (next_node_ready) {
+        ready_queue.push_back(next_node);
+      }
+    }
+    ready_queue.pop_front();
+  }
+  std::set<std::string> outputs;
+  for (const auto& output : config.output()) {
+    auto it = tensors.find(output.name());
+    if (it == tensors.end()) {
+      return tensorflow::errors::InvalidArgument(
+          "must map ensemble output ", output.name(), " for ensemble ",
+          config.name());
+    }
+    if (!it->second.ready) {
+      return tensorflow::errors::InvalidArgument(
+          "no data will be written to ensemble output ", output.name(),
+          " under optimistic assumption for ensemble ", config.name());
     } else {
-      return tensorflow::errors::Internal(
-          "instance group ", group.name(), " of model ", config.name(),
-          " has unexpected kind KIND_AUTO");
+      outputs.insert(it->first);
+    }
+  }
+  // Check redundant ensemble tensors
+  for (const auto& tensor : tensors) {
+    // skip ensemble outputs as they have been check and can have no next nodes
+    if (outputs.find(tensor.first) != outputs.end()) {
+      continue;
+    }
+    if (!tensor.second.ready) {
+      return tensorflow::errors::InvalidArgument(
+          "ensemble tensor ", tensor.first,
+          " is redundant as no data will be written to it under optimistic "
```

Comment:
What is "optimistic assumption"?

Reply:
It means not considering conditionals, just assuming all nodes will be run if the inputs are ready.

Reply:
We don't have any conditionals yet so it makes the error message confusing. The error messages will be important to help users understand why their config is not correct. So write them as accurately and helpfully as possible given the capabilities we have right now.